### PR TITLE
New test framework for testing AssetConditions

### PR DESCRIPTION
**Pull request by OwenKephart**
## Summary & Motivation

As the title -- adds a new test framework that is more ergonomic for testing individual AssetConditions without the overhead of a whole scenario etc.

Mostly as a demonstration, adds a couple of tests for `AssetCondition.missing()`. I'm imagining adding more over time, as well as adding tests for all statically-constructable AssetConditions.

### CODE REVIEW

**Comment by OwenKephart** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/19966?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

Current dependencies on/for this PR:
* **#19966** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/19966?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#19965** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/19965?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#19964** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/19964?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/19966?utm_source=stack-comment).

File: python_modules/dagster/dagster_tests/definitions_tests/auto_materialize_tests/utils/asset_condition_scenario.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+                evaluation_state_by_key={},
+                expected_data_time_mapping={},
+                # If you are an intrepid developer who has added a new field to AssetDaemonContext and
+                # have found yourself here, it's likely that you can just set the value to None and
+                # 'type: ignore' it. Most fields will not be accessed in these tests.
+                daemon_context=AssetDaemonContext(
+                    evaluation_id=1,
+                    instance=self.instance,
+                    asset_graph=self.asset_graph,
+                    cursor=None,  # type: ignore
```
Comment (schrockn) [inline python_modules/dagster/dagster_tests/definitions_tests/auto_materialize_tests/utils/asset_condition_scenario.py:66]: Don't think we should be pumping through objects that violate typing/invariants. Let's construct a dummy AssetDaemonCursor instead.

File: python_modules/dagster/dagster_tests/definitions_tests/auto_materialize_tests/utils/asset_scenario_state.py [hunk 2]
```diff
@@ -155,8 +155,9 @@ def with_all_eager(self, max_materializations_per_minute: int = 1) -> Self:
             )
         )
 
-    def with_current_time(self, time: str) -> Self:
-        return dataclasses.replace(self, current_time=pendulum.parse(time))
+    def with_current_time(self, time: Union[str, datetime.datetime]) -> Self:
```
Comment (schrockn) [inline python_modules/dagster/dagster_tests/definitions_tests/auto_materialize_tests/utils/asset_scenario_state.py:158]: why is this change necessary?

File: python_modules/dagster/dagster_tests/definitions_tests/auto_materialize_tests/utils/asset_condition_scenario.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    asset_condition: Optional[AssetCondition] = None
+    previous_evaluation_state: Optional[AssetConditionEvaluationState] = None
+
+    @staticmethod
+    def create_from_state(
+        state: AssetScenarioState, asset_condition: AssetCondition
+    ) -> "AssetConditionScenarioState":
+        return AssetConditionScenarioState(
+            **{
+                **{f.name: getattr(state, f.name) for f in dataclasses.fields(state)},
```
Comment (schrockn) [inline python_modules/dagster/dagster_tests/definitions_tests/auto_materialize_tests/utils/asset_condition_scenario.py:35]: Not a huge fan of this cleverness (would rather just list these things out which is obvious and you get typechecking)

Also does it ever make sense to create an  `AssetConditionScenarioState` from the daemon variant?

**Review by schrockn — CHANGES_REQUESTED** [review]
Just a few inline comments, plus we might need to rebase depending on resolution to the beginning of this stack.

File: python_modules/dagster/dagster_tests/definitions_tests/auto_materialize_tests/utils/asset_condition_scenario.py (hunk 3 above)
Comment (OwenKephart) [reply python_modules/dagster/dagster_tests/definitions_tests/auto_materialize_tests/utils/asset_condition_scenario.py:35]: It would not really make sense to create an AssetConditionScenarioState from the daemon variant -- this is basically here so we can have a "library" of common setups (dependency relationships etc.) that either subclass can pull from. This is more evidence of it being nicer to go with an approach in which AssetConditionScenarioState contains an AssetScenarioState rather than it being a subclass.

That also gets around this weird field copying approach (which I went with out of mild fear of adding a new field to the parent class and forgetting to add it to this constructor).

File: python_modules/dagster/dagster_tests/definitions_tests/auto_materialize_tests/utils/asset_scenario_state.py (hunk 2 above)
Comment (OwenKephart) [reply python_modules/dagster/dagster_tests/definitions_tests/auto_materialize_tests/utils/asset_scenario_state.py:158]: just QoL -- nice to be able to pass in either a string or a datetime to this sort of method

Comment (OwenKephart) [reply python_modules/dagster/dagster_tests/definitions_tests/auto_materialize_tests/utils/asset_scenario_state.py:158]: (specifically, I do that on line 36 of missing_tests.py)

**Comment by OwenKephart** [comment]
closed in favor of: https://github.com/dagster-io/dagster/pull/20278
